### PR TITLE
bpf: fix asymmetric routing and cilium_host connectivity in v6 tunnel

### LINKED ISSUE #8654: v6 tunneling: cannot ping remote cilium_host ip
(Not related to nodeport, just plain tunneling mode.)

### PR DESCRIPTION
**Pull request by borkmann**
(See commit description. This gets nodeport fully working in v6 tunneling mode.)

Fixes: #8654

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/9062)

### CODE REVIEW

**Comment by borkmann** [comment]
test-me-please

File: bpf/bpf_lxc.c [hunk 1]
```diff
@@ -597,6 +600,8 @@ static inline int handle_ipv4_from_lxc(struct __sk_buff *skb, __u32 *dstID)
 		key.family = ENDPOINT_KEY_IPV4;
 
 		ret = encap_and_redirect_lxc(skb, tunnel_endpoint, encrypt_key, &key, SECLABEL, monitor);
+		if (!revalidate_data(skb, &data, &data_end, &ip4))
+			return DROP_INVALID;
```
Comment (joestringer) [inline bpf/bpf_lxc.c]: The IPv4 path here differs from the IPv6 path in the ordering of the new `revalidate_data()` call vs. encap return value checks. Is this significant?

**Review by joestringer — DISMISSED** [review]
Looks good, one minor nit below which I think we can fix to make the ipv4/ipv6 cases more consistent.

Might need some special logic for encryption mode cases, given the testsuite failure. /cc @jrfastab

File: bpf/lib/encap.h [hunk 2]
```diff
@@ -64,6 +64,63 @@ encap_and_redirect_ipsec(struct __sk_buff *skb, __u32 tunnel_endpoint, __u8 key,
...
+	if (egress) {
+		BPF_V6(host_ip, ROUTER_IP);
+		noff = ETH_HLEN + offsetof(struct ipv6hdr, saddr);
+	} else {
+		BPF_V6(host_ip, HOST_IP);
+		noff = ETH_HLEN + offsetof(struct ipv6hdr, daddr);
+	}
+	sum = csum_diff(which, 16, &host_ip, 16, 0);
+	csum_l4_offset_and_flags(nexthdr, &csum);
+	if (skb_store_bytes(skb, noff, &host_ip, 16, 0) < 0)
```
Comment (jrfastab) [inline bpf/lib/encap.h:115]: would be nice to have some define for '16' number here but not critical could be done later.

**Comment by borkmann** [comment]
#9090 needs to be merged first

**Review by joestringer — APPROVED** [review]

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
Needs 2/10 sec shrinking, seen with sockops + vxlan in CI:
```
2019-08-30T22:41:07.70259549Z level=debug msg="Finished writing ELF" error="<nil>" new-elf-path=1121_next/bpf_lxc.o subsys=elf template-path=/var/run/cilium/state/templates/e32428ce7b33eb2836de11b51ca4def78de19e3f/bpf_lxc.o
2019-08-30T22:41:07.702966258Z level=debug msg="netlink: Replacing qdisc for lxce4c955750fc9 succeeded" subsys=datapath-loader
2019-08-30T22:41:07.712576004Z level=error msg="Command execution failed" cmd="[tc filter replace dev lxce4c955750fc9 ingress prio 1 handle 1 bpf da obj 1121_next/bpf_lxc.o sec from-container]" error="exit status 1" subsys=datapath-loader
2019-08-30T22:41:07.712596182Z level=warning subsys=datapath-loader
2019-08-30T22:41:07.712599839Z level=warning msg="Prog section '2/10' rejected: Invalid argument (22)!" subsys=datapath-loader
2019-08-30T22:41:07.712603616Z level=warning msg=" - Type:         3" subsys=datapath-loader
2019-08-30T22:41:07.712607082Z level=warning msg=" - Attach Type:  0" subsys=datapath-loader
2019-08-30T22:41:07.712610919Z level=warning msg=" - Instructions: 4574 (478 over limit)" subsys=datapath-loader
2019-08-30T22:41:07.712633482Z level=warning msg=" - License:      GPL" subsys=datapath-loader
2019-08-30T22:41:07.712638531Z level=warning subsys=datapath-loader
2019-08-30T22:41:07.712643641Z level=warning msg="Verifier analysis:" subsys=datapath-loader
2019-08-30T22:41:07.712647398Z level=warning subsys=datapath-loader
2019-08-30T22:41:07.712763466Z level=warning msg="Error filling program arrays!" subsys=datapath-loader
2019-08-30T22:41:07.712816276Z level=warning msg="Unable to load program" subsys=datapath-loader
2019-08-30T22:41:07.71417853Z level=warning msg="JoinEP: Failed to load program" containerID=d866a5f8bc datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=1121 error="Failed to load tc filter: exit status 1" file-path=1121_next/bpf_lxc.o identity=46603 ipv4=10.10.1.93 ipv6="f00d::a0a:100:0:2021" k8sPodName=default/testds-4dgnm subsys=datapath-loader veth=lxce4c955750fc9
2019-08-30T22:41:07.714193949Z level=error msg="Error while rewriting endpoint BPF program" containerID=d866a5f8bc datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=1121 error="Failed to load tc filter: exit status 1" identity=46603 ipv4=10.10.1.93 ipv6="f00d::a0a:100:0:2021" k8sPodName=default/testds-4dgnm subsys=endpoint
```

**Review by brb — APPROVED** [review]
LGTM for Go parts.

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-me-please

**Pull request merged**